shaders.go

```diff
@@ -106,193 +107,187 @@ void main (void) {
 	Gl.Enable(Gl.BLEND)
 	Gl.BlendFunc(Gl.SRC_ALPHA, Gl.ONE_MINUS_SRC_ALPHA)
 
-	// We sometimes have to change our projection matrix
-	Mailbox.Listen("WindowResizeMessage", func(m Message) {
-		wrm, ok := m.(WindowResizeMessage)
-		if !ok {
-			return
-		}
+	s.projectionMatrix = make([]float32, 9)
+	s.projectionMatrix[8] = 1
 
-		if !scaleOnResize {
-			s.SetProjection(float32(wrm.NewWidth), float32(wrm.NewHeight))
-		}
-	})
-}
+	s.viewMatrix = make([]float32, 9)
+	s.viewMatrix[0] = 1
+	s.viewMatrix[4] = 1
+	s.viewMatrix[8] = 1
 
-func (s *defaultShader) Pre() {
-	Gl.UseProgram(s.program)
-	Gl.Uniform2f(s.ufProjection, s.projX, s.projY)
-	Gl.Uniform3f(s.ufCamera, cam.x, cam.y, cam.z)
+	s.modelMatrix = make([]float32, 9)
+	s.modelMatrix[0] = 1
+	s.modelMatrix[4] = 1
+	s.modelMatrix[8] = 1
 }
 
-func (s *defaultShader) Draw(texture *gl.Texture, buffer *gl.Buffer, x, y, rotation float32) {
-	if s.lastBuffer != buffer {
-		Gl.BindBuffer(Gl.ARRAY_BUFFER, buffer)
-		Gl.VertexAttribPointer(s.inPosition, 2, Gl.FLOAT, false, 20, 0)
-		Gl.VertexAttribPointer(s.inTexCoords, 2, Gl.FLOAT, false, 20, 8)
-		Gl.VertexAttribPointer(s.inColor, 4, Gl.UNSIGNED_BYTE, true, 20, 16)
+func (s *basicShader) Pre() {
+	Gl.UseProgram(s.program)
 
-		s.lastBuffer = buffer
+	if scaleOnResize {
+		s.projectionMatrix[0] = 1 / (gameWidth / 2)
+		s.projectionMatrix[4] = 1 / (-gameHeight / 2)
+	} else {
+		s.projectionMatrix[0] = 1 / (windowWidth / 2)
+		s.projectionMatrix[4] = 1 / (-windowHeight / 2)
 	}
 
-	if s.lastTexture != texture {
-		Gl.BindTexture(Gl.TEXTURE_2D, texture)
-
-		s.lastTexture = texture
+	if s.cameraEnabled {
+		s.viewMatrix[1], s.viewMatrix[0] = math.Sincos(cam.angle * math.Pi / 180)
+		s.viewMatrix[3] = -s.viewMatrix[1]
+		s.viewMatrix[4] = s.viewMatrix[0]
+		s.viewMatrix[6] = -cam.x
+		s.viewMatrix[7] = -cam.y
+		s.viewMatrix[8] = cam.z
+	} else {
+		s.viewMatrix[6] = -1 / s.projectionMatrix[0]
+		s.viewMatrix[7] = 1 / s.projectionMatrix[4]
```

Comment:
This notation (one line per index) is very verbose, but performs way better than reallocating a new `[]float32{}`.